homeassistant/components/screenlogic/__init__.py

```diff
@@ -0,0 +1,202 @@
+        raise ConfigEntryNotReady from ex
+    except AttributeError as ex:
+        _LOGGER.exception(
+            "Unexpected error while connecting to the gateway %s", connect_info
+        )
+        raise ConfigEntryNotReady from ex
+
+    coordinator = ScreenlogicDataUpdateCoordinator(
+        hass, config_entry=entry, gateway=gateway
+    )
+
+    device_data = defaultdict(list)
+
+    await coordinator.async_refresh()
+
+    for circuit in coordinator.data["circuits"]:
+        device_data["switch"].append(circuit)
+
+    for sensor in coordinator.data["sensors"]:
+        if sensor == "chem_alarm":
+            device_data["binary_sensor"].append(sensor)
+        else:
+            if coordinator.data["sensors"][sensor]["value"] != 0:
+                device_data["sensor"].append(sensor)
+
+    for pump in coordinator.data["pumps"]:
+        if (
+            coordinator.data["pumps"][pump]["data"] != 0
+            and "currentWatts" in coordinator.data["pumps"][pump]
+        ):
+            device_data["pump"].append(pump)
+
+    for body in coordinator.data["bodies"]:
+        device_data["water_heater"].append(body)
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        "coordinator": coordinator,
+        "devices": device_data,
+        "listener": entry.add_update_listener(async_update_listener),
+    }
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    hass.data[DOMAIN][entry.entry_id]["listener"]()
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+async def async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+class ScreenlogicDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage the data update for the Screenlogic component."""
+
+    def __init__(self, hass, *, config_entry, gateway):
+        """Initialize the Screenlogic Data Update Coordinator."""
+        self.config_entry = config_entry
+        self.gateway = gateway
+        self.screenlogic_data = {}
+        interval = timedelta(
+            seconds=config_entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+        )
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=interval,
+        )
+
+    async def _async_update_data(self):
+        """Fetch data from the Screenlogic gateway."""
+        try:
+            self.gateway.update()
```

Comment:
This likely needs to run in the executor as I'm pretty sure it's doing I/o

Reply:
Yup. Resolved with your suggested change, comit [151490f](https://github.com/home-assistant/core/pull/47933/commits/151490fd61078bb086f22f9d644ee38140117d9e).